silkrpc/context_pool.cpp

```diff
@@ -23,83 +23,152 @@
-    asio::detail::thread_group workers{};
-    for (std::size_t i{0}; i < contexts_.size(); ++i) {
-        auto& context = contexts_[i];
-        workers.create_thread([&]() { context.grpc_runner->run(); });
-        SILKRPC_DEBUG << "ContextPool::run context[" << i << "].grpc_runner started: " << &*context.grpc_runner << "\n";
-        workers.create_thread([&]() { context.io_context->run(); });
-        SILKRPC_DEBUG << "ContextPool::run context[" << i << "].io_context started: " << &*context.io_context << "\n";
+ContextPool::~ContextPool() {
+    SILKRPC_TRACE << "ContextPool::~ContextPool started " << this << "\n";
+    stop();
+    SILKRPC_TRACE << "ContextPool::~ContextPool completed " << this << "\n";
+}
+
+void ContextPool::start() {
+    SILKRPC_TRACE << "ContextPool::start started\n";
+
+    if (!stopped_) {
+        // Create a pool of threads to run all of the contexts (each one having 1 threads)
+        for (std::size_t i{0}; i < contexts_.size(); ++i) {
+            auto& context = contexts_[i];
+            context_threads_.create_thread([&, i = i]() {
+                SILKRPC_DEBUG << "thread start context[" << i << "] thread_id: " << std::this_thread::get_id() << "\n";
+                context.execute_loop();
+                SILKRPC_DEBUG << "thread end context[" << i << "] thread_id: " << std::this_thread::get_id() << "\n";
+            });
+            SILKRPC_DEBUG << "ContextPool::start context[" << i << "].io_context started: " << &*context.io_context() << "\n";
+        }
     }
 
-    // Wait for all threads in the pool to exit
-    workers.join();
+    SILKRPC_TRACE << "ContextPool::start completed\n";
+}
+
+void ContextPool::join() {
+    SILKRPC_TRACE << "ContextPool::join started\n";
+
+    // Wait for all threads in the pool to exit.
+    SILKRPC_DEBUG << "ContextPool::join joining...\n";
+    context_threads_.join();
 
-    SILKRPC_DEBUG << "ContextPool::run completed\n";
+    SILKRPC_TRACE << "ContextPool::join completed\n";
 }
 
 void ContextPool::stop() {
     // Explicitly stop all scheduler runnable components
-    SILKRPC_DEBUG << "ContextPool::stop started\n";
+    SILKRPC_TRACE << "ContextPool::stop started\n";
 
     for (std::size_t i{0}; i < contexts_.size(); ++i) {
-        auto& context = contexts_[i];
-        context.io_context->stop();
-        SILKRPC_DEBUG << "ContextPool::stop context[" << i << "].io_context stopped: " << &*context.io_context << "\n";
-        context.grpc_runner->stop();
-        SILKRPC_DEBUG << "ContextPool::stop context[" << i << "].grpc_runner stopped: " << &*context.grpc_runner << "\n";
+        contexts_[i].stop();
+        SILKRPC_DEBUG << "ContextPool::stop context[" << i << "].io_context stopped: " << &*contexts_[i].io_context() << "\n";
     }
-    SILKRPC_DEBUG << "ContextPool::stop completed\n";
+    SILKRPC_TRACE << "ContextPool::stop completed\n";
+}
+
+void ContextPool::run() {
+    start();
+    join();
 }
 
 Context& ContextPool::get_context() {
```

Comment:
could be useful inserts some policy to select context with less queued request or ...

Reply:
I undestand your suggestion, nevertheless I would not add any of them in this PR. The reasons are:
- it is tempting to add smart policies, but simple ones (e.g. counting enqueued requests in each scheduler) are not proven to be correct for real workloads (i.e. incoming requests are different, hence heterogeneous in terms of performance impact)
- some experimentation and measurements for different policies should be done first
- round-robin is a very simple yet effective scheduling policy w/o knowledge of the input distribution

I would prefer for you to open an issue to discuss this feature and, if accepeted, eventually use a dedicated PR for it.

Reply:
Created issue #281